File: android/talk/src/main/kotlin/app/ptt/talk/EncryptedCallTimeline.kt
package app.ptt.talk

import java.time.Instant
import java.util.UUID

internal enum class CallTimelineEventKind(val wire: String) {
    STARTED("started"),
    ANSWERED("answered"),
    PARTICIPANTS_CHANGED("participants_changed"),
    ENDED("ended"),
}

/**
 * A structured call event carried inside the existing pairwise-encrypted chat
 * envelope. The relay only sees ciphertext and the event follows the
 * conversation's normal retention and device-linking rules.
 */
internal data class EncryptedCallTimelineEvent(
    val callId: UUID,
    val kind: CallTimelineEventKind,
    val startedAt: Instant,
    val durationMs: Long = 0,
    val participantCount: Int,
    val endReason: String = "",
)

internal data class EncryptedCallHistoryItem(
    val callId: UUID,
    val channelId: UUID,
    val startedAt: Instant,
    val endedAt: Instant?,
    val durationMs: Long,
    val participantCount: Int,
    val endReason: String,
    val outgoing: Boolean,
    val answeredOnThisAccount: Boolean,
) {
    val missed: Boolean get() = endReason == "missed" && !answeredOnThisAccount
}

internal object EncryptedCallTimelineCodec {
    private const val PREFIX = "ptt-call-event:v1"
    private const val MAX_DURATION_MS = 8 * 60 * 60 * 1_000L
    private val reasonPattern = Regex("[a-z0-9_]{0,32}")

    fun encode(event: EncryptedCallTimelineEvent): String {
        validate(event)
        return listOf(
            PREFIX,
            event.kind.wire,
            event.callId.toString().lowercase(),
            event.startedAt.toEpochMilli().toString(),
            event.durationMs.toString(),
            event.participantCount.toString(),
            event.endReason.ifBlank { "-" },
        ).joinToString("|")
    }

    fun decode(value: String): EncryptedCallTimelineEvent? {
        if (!value.startsWith("$PREFIX|")) return null
        val fields = value.split('|')
        require(fields.size == 7 && fields[0] == PREFIX)
        val event = EncryptedCallTimelineEvent(
            callId = UUID.fromString(fields[2]),
            kind = CallTimelineEventKind.entries.single { it.wire == fields[1] },
            startedAt = Instant.ofEpochMilli(fields[3].toLong()),
            durationMs = fields[4].toLong(),
            participantCount = fields[5].toInt(),
            endReason = fields[6].takeUnless { it == "-" }.orEmpty(),
        )
        validate(event)
        require(fields[2] == event.callId.toString().lowercase())
        return event
    }

    fun history(messages: Collection<ChatMessage>, localAci: String): List<EncryptedCallHistoryItem> =
        messages.mapNotNull { message ->
            runCatching { decode(message.text) }.getOrNull()?.let { event -> message to event }
        }.groupBy { it.second.callId }.mapNotNull { (callId, entries) ->
            val ordered = entries.sortedBy { it.first.sentAt }
            val first = ordered.firstOrNull() ?: return@mapNotNull null
            val latest = ordered.last()
            val started = ordered.firstOrNull { it.second.kind == CallTimelineEventKind.STARTED } ?: first
            val ended = ordered.lastOrNull { it.second.kind == CallTimelineEventKind.ENDED }
            EncryptedCallHistoryItem(
                callId = callId,
                channelId = latest.first.channelId,
                startedAt = started.second.startedAt,
                endedAt = ended?.first?.sentAt,
                durationMs = ended?.second?.durationMs ?: latest.second.durationMs,
                participantCount = ordered.maxOf { it.second.participantCount },
                endReason = ended?.second?.endReason.orEmpty(),
                outgoing = started.first.senderAci.equals(localAci, true),
                answeredOnThisAccount = ordered.any {
                    it.second.kind == CallTimelineEventKind.ANSWERED && it.first.senderAci.equals(localAci, true)
                },
            )
        }.sortedByDescending { it.startedAt }

    private fun validate(event: EncryptedCallTimelineEvent) {
        require(event.startedAt.toEpochMilli() >= 0)
        require(event.durationMs in 0..MAX_DURATION_MS)
        require(event.participantCount in 1..8)
        require(reasonPattern.matches(event.endReason))
        require(event.kind == CallTimelineEventKind.ENDED || event.endReason.isEmpty())
    }
}

File: ios/PttTalk/Sources/PttTalkLib/EncryptedCallTimeline.swift
import Foundation

public enum CallTimelineEventKind: String, Sendable {
    case started
    case answered
    case participantsChanged = "participants_changed"
    case ended
}

/// A structured call event carried inside the existing pairwise-encrypted chat
/// envelope. The relay only sees ciphertext and normal conversation retention
/// and device-linking rules apply.
public struct EncryptedCallTimelineEvent: Equatable, Sendable {
    public let callId: UUID
    public let kind: CallTimelineEventKind
    public let startedAt: Date
    public let durationMs: Int64
    public let participantCount: Int
    public let endReason: String

    public init(
        callId: UUID, kind: CallTimelineEventKind, startedAt: Date,
        durationMs: Int64 = 0, participantCount: Int, endReason: String = ""
    ) {
        self.callId = callId
        self.kind = kind
        self.startedAt = startedAt
        self.durationMs = durationMs
        self.participantCount = participantCount
        self.endReason = endReason
    }
}

public struct EncryptedCallHistoryItem: Equatable, Identifiable, Sendable {
    public let callId: UUID
    public let channelId: UUID
    public let startedAt: Date
    public let endedAt: Date?
    public let durationMs: Int64
    public let participantCount: Int
    public let endReason: String
    public let outgoing: Bool
    public let answeredOnThisAccount: Bool
    public var id: UUID { callId }
    public var missed: Bool { endReason == "missed" && !answeredOnThisAccount }
}

public enum EncryptedCallTimelineCodec {
    private static let prefix = "ptt-call-event:v1"
    private static let maximumDurationMs: Int64 = 8 * 60 * 60 * 1_000

    public static func encode(_ event: EncryptedCallTimelineEvent) throws -> String {
        try validate(event)
        return [
            prefix,
            event.kind.rawValue,
            event.callId.uuidString.lowercased(),
            String(Int64((event.startedAt.timeIntervalSince1970 * 1_000).rounded())),
            String(event.durationMs),
            String(event.participantCount),
            event.endReason.isEmpty ? "-" : event.endReason,
        ].joined(separator: "|")
    }

    public static func decode(_ value: String) throws -> EncryptedCallTimelineEvent? {
        guard value.hasPrefix("\(prefix)|") else { return nil }
        let fields = value.split(separator: "|", omittingEmptySubsequences: false).map(String.init)
        guard fields.count == 7, fields[0] == prefix,
              let kind = CallTimelineEventKind(rawValue: fields[1]),
              let callId = UUID(uuidString: fields[2]),
              callId.uuidString.lowercased() == fields[2],
              let timestamp = Int64(fields[3]),
              let duration = Int64(fields[4]),
              let participantCount = Int(fields[5]) else {
            throw EncryptedChatError.invalidMessage
        }
        let event = EncryptedCallTimelineEvent(
            callId: callId, kind: kind,
            startedAt: Date(timeIntervalSince1970: TimeInterval(timestamp) / 1_000),
            durationMs: duration, participantCount: participantCount,
            endReason: fields[6] == "-" ? "" : fields[6]
        )
        try validate(event)
        return event
    }

    public static func history(messages: [ChatMessage], localAci: String) -> [EncryptedCallHistoryItem] {
        let decoded: [(ChatMessage, EncryptedCallTimelineEvent)] = messages.compactMap { message in
            do {
                guard let event = try decode(message.text) else { return nil }
                return (message, event)
            } catch {
                return nil
            }
        }
        return Dictionary(grouping: decoded, by: { $0.1.callId }).compactMap { callId, entries in
            let ordered = entries.sorted { $0.0.sentAt < $1.0.sentAt }
            guard let first = ordered.first, let latest = ordered.last else { return nil }
            let started = ordered.first(where: { $0.1.kind == .started }) ?? first
            let ended = ordered.last(where: { $0.1.kind == .ended })
            return EncryptedCallHistoryItem(
                callId: callId, channelId: latest.0.channelId,
                startedAt: started.1.startedAt, endedAt: ended?.0.sentAt,
                durationMs: ended?.1.durationMs ?? latest.1.durationMs,
                participantCount: ordered.map { $0.1.participantCount }.max() ?? 1,
                endReason: ended?.1.endReason ?? "",
                outgoing: started.0.senderAci.caseInsensitiveCompare(localAci) == .orderedSame,
                answeredOnThisAccount: ordered.contains {
                    $0.1.kind == .answered && $0.0.senderAci.caseInsensitiveCompare(localAci) == .orderedSame
                }
            )
        }.sorted { $0.startedAt > $1.startedAt }
    }

    private static func validate(_ event: EncryptedCallTimelineEvent) throws {
        let reason = try! NSRegularExpression(pattern: "^[a-z0-9_]{0,32}$")
        let range = NSRange(event.endReason.startIndex..., in: event.endReason)
        guard event.startedAt.timeIntervalSince1970 >= 0,
              (0...maximumDurationMs).contains(event.durationMs),
              (1...8).contains(event.participantCount),
              reason.firstMatch(in: event.endReason, range: range) != nil,
              event.kind == .ended || event.endReason.isEmpty else {
            throw EncryptedChatError.invalidMessage
        }
    }
}
